util/file_storage.py

```diff
@@ -0,0 +1,186 @@
+        ...
+
+    @abstractmethod
+    def get_link(self, obj: FileObject) -> str:
+        """Get a downloadable link for a file object"""
+        ...
+
+    @abstractmethod
+    def delete(self, name: str) -> bool:
+        """Delete a object from the storage
+        :param name: The file name, with the path
+        """
+        ...
+
+
+class S3FileStorage(FileStorage):
+    """S3 specific implementation"""
+
+    BACKEND = "s3"
+    ACL = "public-read"
+    S3_ENDPOINT_URL = "https://{bucket}.s3.{region_code}.amazonaws.com"
+    BUCKET_REGION_CACHE = {}
+
+    def __init__(self) -> None:
+        config = Configuration.aws_config()
+        extras = dict(endpoint_url=config.endpoint_url)
+
+        # The client changes if we are using a profile name or not
+        if config.profile_name:
+            session = boto3.Session(profile_name=config.profile_name)
+            self.client = session.client("s3", **extras)
+        else:
+            # If the key_id and secret_key are None,
+            # the client should use the 'default' profile automatically
+            self.client = boto3.client(
+                "s3",
+                aws_access_key_id=config.key_id,
+                aws_secret_access_key=config.secret_key,
+                **extras,
+            )
+        self._bucket_name = config.bucket_name
+
+        # The web url to the bucket is required so we may create unsigned public urls
+        if config.endpoint_url is None:
+            if self._bucket_name not in self.BUCKET_REGION_CACHE:
+                region = self.client.get_bucket_location(Bucket=self._bucket_name)
+                self.BUCKET_REGION_CACHE[self._bucket_name] = region[
+                    "LocationConstraint"
+                ]
+
+            region_code = self.BUCKET_REGION_CACHE[self._bucket_name]
+            self.bucket_url = self.S3_ENDPOINT_URL.format(
+                bucket=self._bucket_name, region_code=region_code
+            )
+        else:
+            self.bucket_url = f"{config.endpoint_url}/{self._bucket_name}"
+
+    def write(
+        self, name: str, io: IO, content_type="binary/octet-stream"
+    ) -> Optional[FileObject]:
+        response = self.client.put_object(
+            Key=name,
+            Bucket=self._bucket_name,
+            Body=io.read(),
+            ACL=self.ACL,
+            ContentType=content_type,
+        )
+        if response.get("ResponseMetadata", {}).get("HTTPStatusCode") == 200:
+            return FileObject(
+                key=name, container=self._bucket_name, backend=self.BACKEND
+            )
+        return None
+
+    def delete(self, name: str) -> bool:
+        response = self.client.delete_object(Key=name, Bucket=self._bucket_name)
+        return response.get("ResponseMetadata", {}).get("HTTPStatusCode") == 204
+
+    def get_link(self, obj: FileObject) -> str:
+        """All objects are public-read, return the path to the object"""
+        return f"{self.bucket_url}/{obj.key}"
```

Comment:
Instead of writing our own functions to generate a s3 URL, I think we can simplify this a bit by just use the boto 3 functions, which may be a bit more resilient if aws makes some URL creation changes. The boto3 function naming is a bit silly because we don't actually want presigned URLs, but I think if we call `generate_presigned_url` and its configured with `botocore.UNSIGNED` we should just get a plain unsigned URL. 

https://github.com/boto/boto3/issues/1415

Reply:
Didn't know about this, Thanks!